naga/src/proc/overloads/rule.rs

```diff
@@ -0,0 +1,175 @@
+/*! Type rules.
+
+An implementation of [`OverloadSet`] represents a set of type rules, each of
+which has a list of types for its arguments, and a conclusion about the
+type of the expression as a whole.
+
+This module defines the [`Rule`] type, representing a type rule from an
+[`OverloadSet`], and the [`Conclusion`] type, a specialized enum for
+representing a type rule's conclusion.
+
+[`OverloadSet`]: super::OverloadSet
+
+*/
+
+use super::constructor_set::ConstructorSize;
+
+use crate::common::DiagnosticDebug;
+use crate::ir;
+use crate::proc::TypeResolution;
+use crate::UniqueArena;
+
+use alloc::vec::Vec;
+use core::fmt;
+use core::result::Result;
+
+/// A single type rule.
+#[derive(Clone)]
+pub struct Rule {
+    pub arguments: Vec<TypeResolution>,
+    pub conclusion: Conclusion,
+}
+
+/// The result type produced by a Naga IR builtin function.
+///
+/// A `Conclusion` value represents the return type of some operation
+/// in the builtin function database.
```

Comment:
nitpick: This feels like it could be more brief, but with the same information and no loss to readability.

suggestion: It would also be helpful here to link to the function database. I suppose this would be a link to the `overloads` module?

Combined suggestions:

```suggestion
/// A return type produced by a Naga IR builtin function,
/// generally queried from the [builtin function
/// database](crate::proc::overloads).
```

Reply:
I 100% agree. I must have written this while tired.